File: C-Plus-Plus/other/First_Missing_Positive_Number.cpp
/*
Given an array of integers, find the first missing positive integer in linear time and constant space. In other words, 
find the lowest positive integer that does not exist in the array. The array can contain duplicates and negative numbers as well.

For example, the input [3, 4, -1, 1] should give 2. The input [1, 2, 0] should give 3.

You can modify the input array in-place.

*/

#include<bits/stdc++.h>

using namespace std;

// finding the first missing positive number
int firstMissingNumber(int arr[], int n)
{
    int index, temp;
    // traversing the array
    for (index = 0; index < n; index++)
    {
        // We will ignore the negative numbers and the numbers greater than the array size
        // Also, check if the elements are in correct position, if they is not then do swap
        // operation
        while (arr[index] != arr[arr[index] - 1] && arr[index] >= 1 && arr[index] <= n)
        {
            // Swap operation
            temp = arr[arr[index] - 1];
            arr[arr[index] - 1] = arr[index];
            arr[index] = temp;
        }
    }
    // traversing the array one more time
    for (index = 0; index < n; index++)
    {
        // checking if all the elements are in the correct position
        // The one which is not in the correct position is the first 
        // missing positive number
        if ((index + 1) != arr[index])
            return index + 1;
    }
    return n + 1;
}

// DRIVER FUNCTION
int main()
{
    // size of the array
    int n;
    cout<<"Enter the size of the array: \n";
    // getting the size of the array
    cin >> n;
    int arr[n], index, res;
    cout<<"Enter the elements of the array: \n";
    // getting the array elements
    for (index = 0; index < n; index++)
        cin >> arr[index];
    printf("\n");
    // finding the first positive missing number
    res = firstMissingNumber(arr, n);
    cout <<"The First Missing Positive Number in the given array is: "<<res;
    return 0;
}
/*
For ex:
Example1:
Input:
4
3 4 7 1

Output:
2

Example 2:
Input:
4
3 4 -1 1

Output:
2
*/

File: Java/other/FirstMissingPositiveNumber.java
import java.util.*;

public class FirstMissingPositiveNumber {

    private static void swap(int arr[], int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    private static int findMissingPositive(int arr[]) {
        // we firstly separate the positive numbers from negative numbers
        int j = 0, i;
        // j is for counting non-positive integers
        for (i = 0; i < arr.length; i++) {
            if (arr[i] <= 0) {
                swap(arr, i, j);
                j += 1;
            }
        }

        // getting array for only positive numbers in arr2
        int arr2[] = new int[arr.length - j];
        i = j;
        j = 0;
        for (; i < arr.length; i++) {
            arr2[j] = arr[i];
            j += 1;
        }

        // marking the number visited by its negative
        // for only those that don't go beyond length of the
        // array of positive numbers
        for (i = 0; i < arr2.length; i++) {
            int checkNum = Math.abs(arr2[i]);
            if (checkNum - 1 < arr2.length && arr2[checkNum - 1] > 0) {
                arr2[checkNum - 1] = -arr2[checkNum - 1];
            }
        }

        for (i = 0; i < arr2.length; i++) {
            if (arr2[i] > 0) {
                return i + 1;
            }
        }
        return arr2.length + 1;
    }
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        //taking input array
        System.out.println("Enter size of array:");
        int size = sc.nextInt();
        int arr[] = new int[size];
        System.out.println("Enter array elements:");
        for (int i=0; i<size; i++) {
            arr[i] = sc.nextInt();
        }
        sc.close();

        int missingNum = findMissingPositive(arr);

        System.out.println("Lowest missing positive integer in the array is:" + missingNum);
    }
}
/**
 * Sample input/output:
 * Enter size of array:
 * 8
 * Enter array elements:
 * 1
 * -1
 * -5
 * -3
 * 3
 * 4
 * 2
 * 8
 * Lowest missing positive integer in the array is:5
 * 
 * Time complexity: O(n)
 * Space complexity: O(1)
 */
